R/bm-read-json.R

```diff
@@ -0,0 +1,107 @@
+read_json <- Benchmark(
+  "read_json",
+  setup = function(source = names(known_sources),
+                   reader = c("arrow", "jsonlite", "ndjson", "RcppSimdJson"),
+                   compression = c("uncompressed", "gzip"),
+                   output_format = c("arrow_table", "data_frame")) {
+    reader <- match.arg(reader)
+    compression <- match.arg(compression)
+    output_format <- match.arg(output_format)
+
+    input_file <- ensure_format(source, "json", compression)
+
+    BenchEnvironment(
+      # Map string param name to function
+      read_func = get_json_reader(reader),
+      input_file = input_file,
+      result_dim = get_source_attr(source, "dim"),
+      as_data_frame = output_format == "data_frame"
+    )
+  },
+  before_each = {
+    result <- NULL
+  },
+  run = {
+    result <- read_func(input_file, as_data_frame = as_data_frame)
+  },
+  after_each = {
+    result_class <- if (as_data_frame) "data.frame" else c("Table", "ArrowObject")
+    correct_format <- inherits(result, result_class)
+
+    # TODO revert to error once null types are removed from sources
+    if (ncol(result) != result_dim[2]) {
+      warning("The number of columns do not match; null types dropped or data missing", call. = FALSE)
+    }
+    stopifnot(
+      # "The dimensions do not match" = all.equal(dim(result), result_dim),
+      "The number of rows does not match" = nrow(result) == result_dim[1],
+      "The format isn't correct" = correct_format
+    )
+    result <- NULL
+  },
+  valid_params = function(params) {
+    drop <- ( params$output_format == "arrow_table" & params$reader != "arrow" ) |
+      ( params$reader == "jsonify" & params$compression == "gzip")
+    params[!drop, ]
+  },
+  packages_used = function(params) {
+    if ('RcppSimdJson' %in% params$reader) {
+      c(params$reader, 'data.table')
+    } else {
+      params$reader
+    }
+  }
+)
+
+#' Get a JSON reader
+#'
+#' @param reader string of the reader package to use
+#'
+#' @return the JSON function
+#'
+#' @export
+get_json_reader <- function(reader) {
+  get_con <- function(path) {
+    if (endsWith(path, ".gz")) {
+      con <- gzfile(path, open = "rb")
+    } else {
+      con <- file(path, open = "r")
+    }
+    con
+  }
+
+  reader_functions <- list(
+    arrow = arrow::read_json_arrow,
+    jsonlite = function(path, ...) {
+      con <- get_con(path)
+      on.exit(close(con))
+      jsonlite::stream_in(con = con, verbose = FALSE)
+    },
+    ndjson = function(..., as_data_frame) ndjson::stream_in(...),
+    # NOTE: Removed jsonify from default options because it's unstable
+    jsonify = function(path, ..., as_data_frame) {
+      con <- get_con(path)
+      on.exit(close(con))
+      jsonify::from_ndjson(con, ...)
+    },
+    RcppSimdJson = function(path, ..., as_data_frame) {
+      lines <- readLines(path)
+      json <- RcppSimdJson::fparse(lines, single_null = NA, max_simplify_lvl = 3L, ...)
+      data.table::rbindlist(json)
```

Comment:
I'm fine with keeping this here, though curious if there are other ways we might want to run this an rbind with either base or dplyr versions as well and if we should parameterize those (even if we don't test all of them all the time)

Reply:
The other packages handle rbinding to dataframe automatically (well, you can turn it off for {jsonlite} and do it manually if you want), so this is the only one that would get parameterized. {data.table} felt...appropriate? as also popular on the R-finance side and won't be arguable as not optimized enough. If we want we could parameterize for {RcppSimdJson} and {jsonlite} and I am curious what proportion of time is rbinding, but because {arrow} doesn't support not simplifying I wasn't sure to what extent we care. 

Also we have to be careful with `do.call(rbind, <ndjson list>)` because some of these do support implicit nulls (where the field just doesn't exist), which the base solution will screw up because of recycling. I only used explicit nulls in writing because it seems like a best practice anyway, but there's space here to screw it up if you pick the wrong combo. And arguably the lower level of robustness means the base solution is not quite equivalent.

Reply:
> If we want we could parameterize for {RcppSimdJson} and {jsonlite} and I am curious what proportion of time is rbinding, but because {arrow} doesn't support not simplifying I wasn't sure to what extent we care.

Yeah, we probably don't care _that much_ ultimately — though being able to parameterize would tell us if {data.table} is fastest (I assume it is in this case, but it would be good to measure them all and check that). We can totally punt that until a later ticket too — it's a bit orthogonal to an extension anyway

Reply:
I thought this would be pretty simple to add, but it turned out to be....a little not. The results are interesting, though; `dplyr::bind_rows()` is surprisingly slow. The base rbinding for {jsonlite} even includes an extra 

```r
lapply(out, function(x) lapply(x, function(y) if (is.null(y)) NA else y))
```
to replace `NULL` with `NA` (which it can't do on read, annoyingly, and `do.call(rbind.data.frame, ...)` won't work with NULLs), and it's _still_ faster than `dplyr::bind_rows()`. 

Also {jsonlite}-native simplification is faster than `data.table::rbindlist()`...maybe it can make stronger assumptions? Or maybe there's a lot of simplification overhead in {jsonlite} even if you try to turn it off? Not sure.

Using `type_integers`, uncompressed:

```tsv
   process   real gc_level0 gc_level1 gc_level2 reader       output_format rbinder    cpu_count
     <dbl>  <dbl>     <int>     <int>     <int> <chr>        <chr>         <chr>          <int>
 1   0.726  0.106         0         0         0 arrow        arrow_table   package           10
 2   0.775  0.108         0         0         0 arrow        data_frame    package           10
 3   4.62   4.55          3         3         7 RcppSimdJson data_frame    data.table        10
 4   6.90   6.81          4         2         7 ndjson       data_frame    package           10
 5   8.31   8.23         24         8         9 jsonlite     data_frame    package           10
 6  12.0   11.8          27         4         8 RcppSimdJson data_frame    base              10
 7  18.3   18.1         115        12        10 jsonlite     data_frame    data.table        10
 8  19.6   19.3          43         8        12 RcppSimdJson data_frame    dplyr             10
 9  31.1   30.6         143        17        13 jsonlite     data_frame    base              10
10  45.3   44.9         171        21        19 jsonlite     data_frame    dplyr             10
```

Reply:
We could also try swapping out `readLines()` and `readr::read_lines()` or `vroom::vroom_lines()` for {RcppSimdJson} (and {jsonlite} if we like) if we want to try to give them the best chance possible, though it will make much less of a difference.

Also FWIW noticed in `?ndjson::stream_in()`:

> the package uses `data.table::rbindlist` for speed

Reply:
Interesting, indeed!